abci/example/kvstore/kvstore.go

```diff
@@ -170,3 +170,18 @@ func (app *Application) Query(reqQuery types.RequestQuery) (resQuery types.Respo
 
 	return resQuery
 }
+
+func (app *Application) PrepareProposal(req types.RequestPrepareProposal) types.ResponsePrepareProposal {
+	return types.ResponsePrepareProposal{
+		BlockData: req.BlockData,
+	}
+}
+
+func (app *Application) ProcessProposal(req types.RequestProcessProposal) types.ResponseProcessProposal {
+	for _, tx := range req.BlockData.Txs {
+		if len(tx) == 0 {
```

Comment:
This is a bug that we also identified in `v0.34.x`.
It is a very graphical example of violation of `PrepareProposal`/`ProcessProposal` coherence (see Requirement 3 [here](https://github.com/tendermint/tendermint/blob/main/spec/abci/abci%2B%2B_app_requirements.md#formal-requirements)).
More in detail, `PrepareProposal` accepts empty transactions but `ProcessProposal` rejects them. Therefore, if a client submits an empty transaction to a node, that node will propagate the transaction to all validators, which will always propose blocks with that transaction, rejected by `ProcessProposal`. End result: the chain will halt.
This is precisely the concern that drives [this issue](https://github.com/informalsystems/audit-celestia/issues/33).